orchestrator/test_runner/src/eip_712.rs

```diff
@@ -0,0 +1,134 @@
+    let receiver =
+        CosmosAddress::from_bech32(RECEIVER.to_string()).expect("invalid successful receiver?");
+    // Grant the user the tokens needed for the message to succeed
+    setup(contact, &keys, sender).await;
+
+    let http_client = Client::new();
+    // Get the URL from contact, remove the port, add the tx POST port
+    let tx_post_host =
+        contact.get_url().rsplit_once(':').unwrap().0.to_string() + ":" + TX_POST_PORT;
+    let tx_post_url = tx_post_host + TX_POST_ENDPOINT;
+    info!("Sending Tx to {tx_post_url}");
+    let req_body = wrap_transaction(SUCCESSFUL_EIP712_MSG_SEND);
+    info!("Tx request is {req_body}");
+
+    let sender_prebal = contact
+        .get_balance(sender, STAKING_TOKEN.to_string())
+        .await
+        .expect("Could not get sender's balance")
+        .expect("Sender has no balance");
+    let receiver_prebal = contact
+        .get_balance(receiver, STAKING_TOKEN.to_string())
+        .await
+        .expect("Could not get receiver's balance")
+        .expect("receiver has no balance");
+
+    // Send the transaction
+    let res = http_client
+        .post(tx_post_url.clone())
+        .header(reqwest::header::CONTENT_TYPE, "application/json")
+        .body(req_body)
+        .send()
+        .await
+        .expect("Transaction submission failure")
+        .text()
+        .await;
+    debug!("Tx submission res: {res:?}");
+
+    let sender_postbal = contact
+        .get_balance(sender, STAKING_TOKEN.to_string())
+        .await
+        .expect("Could not get sender's balance")
+        .expect("Sender has no balance");
+    let receiver_postbal = contact
+        .get_balance(receiver, STAKING_TOKEN.to_string())
+        .await
+        .expect("Could not get receiver's balance")
+        .expect("receiver has no balance");
+
+    let transferred: Uint256 = AMOUNT.into();
+    let total_paid: Uint256 = transferred + FEE.into();
+
+    assert!(sender_postbal.amount < sender_prebal.amount);
+    let sender_diff = sender_prebal.amount - sender_postbal.amount;
+    assert_eq!(sender_diff, total_paid);
+    assert!(receiver_postbal.amount > receiver_prebal.amount);
+    let receiver_diff = receiver_postbal.amount - receiver_prebal.amount;
+    assert_eq!(receiver_diff, transferred);
+
+    let second_req = wrap_transaction(BAD_SECOND_MESSAGE);
+    let res = http_client
+        .post(tx_post_url)
+        .header(reqwest::header::CONTENT_TYPE, "application/json")
+        .body(second_req)
+        .send()
+        .await
+        .expect("Transaction submission failure")
+        .text()
+        .await;
+    debug!("Expected failure res: {res:?}");
+    assert!(res.is_ok());
+    let res = res.unwrap();
+    assert!(res.contains("signature verification failed")); // The signature must be invalid
+    assert!(res.contains("feePayer pubkey EthPubKeySecp256k1{026016DDBE4183D2FE98CBDB0EDC321F42FACF24F6388B2775C078C7FDBD59FE02} is different from transaction pubkey"));
```

Comment:
can we confirm that the balance has not changed here? what if the error message was printed but the tx accepted.

Reply:
Also in this test we should build a deep space tx, mess up it's signature and then try to submit it, just to also be sure there.